cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -390,6 +390,238 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             ));
         }
 
+        Opcode::Popcnt => {
```

Comment:
I'm sure this works correctly, but it concerns me in two ways:

* The text here is long and it's (almost) exactly a 6-way unrolled loop in the 64-bit case.  Would it be possible to do this by having a loop here (in lower.rs, not the generated code) which reads a table of constants?  In the style of https://sourceware.org/git/?p=valgrind.git;a=blob;f=VEX/priv/guest_amd64_toIR.c;h=fadf47d41d6ba6f63a5ca3800409b11da94c5f7e;hb=HEAD#l4948

* I see an integer multiply at the end.  In the sequences I'm familiar with, that doesn't appear.  It's probably a 3-cycle latency, compared with the 1-cycle latency and parallelisable over #-available-basic-ALUs for a sequence using only and/or/add/sub/shifts.

Given that the current goal is to get correct x64 code generation landed, I'd be happy for these to be looked at in a followup.  They don't need to block landing.

Reply:
I've mostly ported what the old backend did, and I'm happy to revisit it later when this has landed. Thanks for the comments!